src/irmin/tree_intf.ml

```diff
@@ -326,20 +326,23 @@ module type S = sig
   (** {1 Proofs} *)
 
   module Proof : sig
-    type tree
-    type t = (hash, step, metadata) Proof.t [@@deriving irmin]
+    include
+      Proof.S
+        with type contents := contents
+         and type hash := hash
+         and type step := step
+         and type metadata := metadata
+         and type proof = (contents, hash, step, metadata) Proof.proof
 
-    val of_tree : tree -> t
-    (** [of_tree t] is the proof representing the tree [t]. Shallow hashes will
-        be blinded. *)
+    type tree
 
-    val to_tree : t -> tree
-    (** [of_proof p] is the tree representing the proof [p]. Blinded parts of
-        the proof will raise [Dangling_hash] when traversed. *)
+    val of_f : tree -> (tree -> tree Lwt.t) -> proof Lwt.t
```

Comment:
Is that `of_f` really necessary? It will leak the cache out of `f` through the first parameter

Reply:
yea I am not super convinced as well but otherwise, it was hard to port the tests :-) I could probably just change the tests instead.

Reply:
Removed now